### PR TITLE
Add GRPC Rust library target that is based on Tonic

### PR DESCRIPTION
**Pull request by lukas-slezevicius**
## What is the goal of this PR?

Add `//grpc/rust:typedb-protocol` target that is a `rust_library` that contains TypeDB Protocol's automatically generated protobuf code. The library used for this is `tonic`, which we are planning to use instead of `grpc` as it's more actively developed, popular, and mature.

## What are the changes implemented in this PR?

* Create `filegroup` targets for each directory that contains protobuf
* Update `vaticle_dependencies` to include `tonic` and related crates
* Add the targets required for creating the `rust_library`
* Add `lib.rs` to include the generated code
* Add `build.rs` to compile protobuf during Rust's compilation stage

### CODE REVIEW

**Comment by grabl** [comment]
### PR Review Checklist

Do not edit the content of this comment. The PR reviewer should simply update this comment by ticking each review item below, as they get completed.

----
**Code**

- [ ] Packages, classes, and methods have a single domain of responsibility.
- [ ] Packages, classes, and methods are grouped into cohesive and consistent domain model.
- [ ] The code is canonical and the minimum required to achieve the goal.
- [ ] Modules, libraries, and APIs are easy to use, robust (foolproof and not errorprone), and tested.
- [ ] Logic and naming has clear narrative that communicates the accurate intent and responsibility of each module (e.g. method, class, etc.).
- [ ] The code is algorithmically efficient and scalable for the whole application.

**Architecture**

- [ ] Any required refactoring is completed, and the architecture does not introduce technical debt incidentally.
- [ ] Any required build and release automations are updated and/or implemented.
- [ ] Any new components follows a consistent style with respect to the pre-existing codebase.
- [ ] The architecture intuitively reflects the application domain, and is easy to understand.
- [ ] The architecture has a well-defined hierarchy of encapsulated components.
- [ ] The architecture is extensible and scalable.

File: grpc/rust/BUILD [hunk 1]
```diff
@@ -85,3 +86,34 @@ checkstyle_test(
...
+    name = "build_script",
+    srcs = ["build.rs"],
+    data = [
+        "//cluster:proto-files",
+        "//common:proto-files",
+        "//core:proto-files",
+        "@com_google_protobuf//:protoc",
+    ],
+    build_script_env = {
+        "PROTOC": "$(execpath @com_google_protobuf//:protoc)"
```
Comment (lukas-slezevicius) [inline grpc/rust/BUILD:100]: Uses protoc provided by bazel.

File: grpc/rust/BUILD [hunk 2]
```diff
@@ -85,3 +86,34 @@ checkstyle_test(
...
     size = "small",
 )
+
+cargo_build_script(
+    name = "build_script",
+    srcs = ["build.rs"],
+    data = [
+        "//cluster:proto-files",
+        "//common:proto-files",
+        "//core:proto-files",
```
Comment (lukas-slezevicius) [inline grpc/rust/BUILD:96]: Passing the paths of these targets through `rootpath` did not work, so I left it as it is.

File: grpc/rust/lib.rs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+//
+
+tonic::include_proto!("typedb.protocol");
```
Comment (alexjpwalker) [inline grpc/rust/lib.rs:18]: Why do we use a one-line file here? Is there some way we can merge `lib.rs` and `build.rs` - or is there a specific reason why having both is useful?

File: cluster/BUILD [hunk 4]
```diff
@@ -50,3 +50,8 @@ checkstyle_test(
     license_type = "agpl-header",
     size = "small",
 )
+
+filegroup(
```
[the remark this answers is not in the document]
Comment (lukas-slezevicius) [reply cluster/BUILD]: Implemented the changes.

File: grpc/rust/lib.rs (hunk 3 above)
Comment (lukas-slezevicius) [reply grpc/rust/lib.rs:18]: `rust_library` requires at least a single source file that would contain the library's code. The one-line macro simply includes all the code generated through `protoc`. This way, you can just import the `rust_library` from your code directly.

`build.rs` is the build file that runs before compiling the `rust_library`.

You can't join the files, as the build script runs before compiling the library, and the library relies on the output of the build script.

File: grpc/rust/BUILD [hunk 5]
```diff
@@ -85,3 +86,44 @@ checkstyle_test(
...
+    build_script_env = {
+        "PROTOC": "$(execpath @com_google_protobuf//:protoc)"
+    },
+    deps = [
+        "@vaticle_dependencies//library/crates:tonic_build",
+    ],
+)
+
+rust_library(
+    name = "lib",
```
Comment (alexjpwalker) [inline grpc/rust/BUILD:118]: Let's rename this to `typedb-protocol`, for consistency with Java, Python and Node.js.

This is in line with our other library repos, where the "entry point" - i.e the principal export of the repo - is named the same as the repo.

Comment (lukas-slezevicius) [reply grpc/rust/BUILD]: Done.

**Review by alexjpwalker — APPROVED** [review]

**Review by haikalpribadi — APPROVED** [review]
1 change to make before merging.

**Pull request merged**